mlir/lib/Catalyst/Transforms/ApplyTransformSequencePass.cpp

```diff
@@ -0,0 +1,93 @@
+// Copyright 2024 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// This pass applies the passes scheduled with the transform dialect,
+// and then removes the transformer module from the payload.
+
+#define DEBUG_TYPE "applytransformsequence"
+
+#include "Catalyst/IR/CatalystDialect.h"
+#include "mlir/Dialect/Transform/Interfaces/TransformInterfaces.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/Pass/Pass.h"
+#include "llvm/Support/Debug.h"
+
+using namespace llvm;
+using namespace mlir;
+using namespace catalyst;
+
+namespace catalyst {
+#define GEN_PASS_DEF_APPLYTRANSFORMSEQUENCEPASS
+#include "Catalyst/Transforms/Passes.h.inc"
+
+struct ApplyTransformSequencePass
+    : public impl::ApplyTransformSequencePassBase<ApplyTransformSequencePass> {
+    using impl::ApplyTransformSequencePassBase<
+        ApplyTransformSequencePass>::ApplyTransformSequencePassBase;
+
+    void runOnOperation() override
+    {
+        // The top-level module is the payload
+        // We need to remove the transformer module from the payload,
+        // without deleting the transformer module in memory.
+        // Then, apply the transformer module to the payload.
+
+        Operation *payload = getOperation();
+        Operation *transformer;
+
+        // Find the transformer module and remove it from payload
+        payload->walk([&](Operation *op) {
+            if (isa<ModuleOp>(op)) {
+                if (op->hasAttr("transform.with_named_sequence")) {
+                    transformer = op;
+                }
+            }
+        });
```

Comment:
Look a little bit more into the walkers. You can do several things to speed them up, simplify them. The first one is running only when you find a nested ModuleOp. The second one is breaking the iteration. See WalkResult::interrupt(); and ::skip.

Reply:
Looking at this comment https://github.com/llvm/llvm-project/blob/ca409892c5396fa3fbb8ea4dbf53d0e952f36d09/mlir/examples/transform-opt/mlir-transform-opt.cpp#L259-L264

```
  // Identify the module containing the transform script entry point. This may
  // be the same module as the input or a separate module. In the former case,
  // make a copy of the module so it can be modified freely. Modification may
  // happen in the script itself (at which point it could be rewriting itself
  // during interpretation, leading to tricky memory errors) or by embedding
  // library modules in the script.
```

I would implement something similar.

Reply:
Yes, I actually thought about how to break from the visitor once the transformer module node is found but I did not know about the interrupt and skip. I will try them out!

Reply:
done